### PR TITLE
Vladimir Jeremic - UI Design Patterns Challenge

### PR DESCRIPTION
**Pull request by xvladax**

### CODE REVIEW

File: MVP Module/Tests/Feed UI/FeedUIIntegrationTests.swift [hunk 1]
```diff
@@ -280,6 +280,51 @@ final class FeedUIIntegrationTests: XCTestCase {
...
+		let (sut, loader) = makeSUT()
+
+		sut.loadViewIfNeeded()
+		loader.completeFeedLoadingWithError()
+		XCTAssertNotNil(sut.errorMessage, "Expected to show error messagge after failed load")
+
+		sut.simulateUserInitiatedFeedReload()
+		loader.completeFeedLoadingWithError()
+		XCTAssertNotNil(sut.errorMessage, "Expected to show error messagge after failed reload")
+	}
```
Comment (caiozullo) [inline MVP Module/Tests/Feed UI/FeedUIIntegrationTests.swift:315]: These three tests can be combined into a single one:

```swift
func test_loadFeedCompletionWithError_showsErrorMessageUntilNextReload() {
    let (sut, loader) = makeSUT()

    sut.loadViewIfNeeded()
    XCTAssertEqual(sut.errorMessage, nil)

    loader.completeFeedLoadingWithError()
    XCTAssertEqual(sut.errorMessage, localized("FEED_VIEW_CONNECTION_ERROR"))

    sut.simulateUserInitiatedFeedReload()
    XCTAssertEqual(sut.errorMessage, nil)
}
```

File: MVP Module/Tests/Feed UI/FeedUIIntegrationTests.swift [hunk 2]
```diff
@@ -280,6 +280,51 @@ final class FeedUIIntegrationTests: XCTestCase {
...
+	func test_loadFeedError_hidesErrorMessageOnUserTap() {
+		let (sut, loader) = makeSUT()
+
+		sut.loadViewIfNeeded()
+		loader.completeFeedLoadingWithError()
+		XCTAssertNotNil(sut.errorMessage, "Expected to show error messagge after failed load")
+
+		sut.simulateTapOnErrorMessage()
+		XCTAssertNil(sut.errorMessage, "Expected to hide error messagge after user tap action")
+	}
```
Comment (caiozullo) [inline MVP Module/Tests/Feed UI/FeedUIIntegrationTests.swift:326]: This test name and assertions can be improved to clarify intent:

```swift
func test_errorView_hidesErrorMessageOnTap() {
    let (sut, loader) = makeSUT()

    sut.loadViewIfNeeded()
    XCTAssertEqual(sut.errorMessage, nil)

    loader.completeFeedLoadingWithError()
    XCTAssertEqual(sut.errorMessage, localized("FEED_VIEW_CONNECTION_ERROR"))

    sut.simulateTapOnErrorMessage()
    XCTAssertEqual(sut.errorMessage, nil)
}
```

File: MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 3]
```diff
@@ -27,11 +29,20 @@ public final class FeedViewController: UITableViewController, UITableViewDataSou
 		title = viewModel?.title
 		viewModel?.onLoadingStateChange = { [weak self] isLoading in
 			if isLoading {
+				self?.errorView?.hideMessage()
```
Comment (caiozullo) [inline MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift:32]: The ViewModel decides when to hide/show errors by calling `onLoadError`. So this line can be removed.

```suggestion
```

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift [hunk 4]
```diff
@@ -20,13 +20,17 @@ final class FeedViewModel {
 
 	var onLoadingStateChange: Observer<Bool>?
 	var onFeedLoad: Observer<[FeedImage]>?
+	var onLoadError: Observer<String?>?
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/FeedViewModel.swift:23]: `onLoadError` could be renamed to `onLoadingStateChange` since it's used to hide errors as well (with `nil`).

File: MVVM Module/Tests/Feed UI/FeedUIIntegrationTests.swift [hunk 5]
```diff
@@ -280,6 +280,51 @@ final class FeedUIIntegrationTests: XCTestCase {
...
+		let (sut, loader) = makeSUT()
+
+		sut.loadViewIfNeeded()
+		loader.completeFeedLoadingWithError()
+		XCTAssertNotNil(sut.errorMessage, "Expected to show error messagge after failed load")
+
+		sut.simulateUserInitiatedFeedReload()
+		loader.completeFeedLoadingWithError()
+		XCTAssertNotNil(sut.errorMessage, "Expected to show error messagge after failed reload")
+	}
```
Comment (caiozullo) [inline MVVM Module/Tests/Feed UI/FeedUIIntegrationTests.swift:315]: These three tests can be combined into a single one:

```swift
func test_loadFeedCompletionWithError_showsErrorMessageUntilNextReload() {
    let (sut, loader) = makeSUT()

    sut.loadViewIfNeeded()
    XCTAssertEqual(sut.errorMessage, nil)

    loader.completeFeedLoadingWithError()
    XCTAssertEqual(sut.errorMessage, localized("FEED_VIEW_CONNECTION_ERROR"))

    sut.simulateUserInitiatedFeedReload()
    XCTAssertEqual(sut.errorMessage, nil)
}
```

File: MVVM Module/Tests/Feed UI/FeedUIIntegrationTests.swift [hunk 6]
```diff
@@ -280,6 +280,51 @@ final class FeedUIIntegrationTests: XCTestCase {
...
+	func test_loadFeedError_hidesErrorMessageOnUserTap() {
+		let (sut, loader) = makeSUT()
+
+		sut.loadViewIfNeeded()
+		loader.completeFeedLoadingWithError()
+		XCTAssertNotNil(sut.errorMessage, "Expected to show error messagge after failed load")
+
+		sut.simulateTapOnErrorMessage()
+		XCTAssertNil(sut.errorMessage, "Expected to hide error messagge after user tap action")
+	}
```
Comment (caiozullo) [inline MVVM Module/Tests/Feed UI/FeedUIIntegrationTests.swift:326]: This test name and assertions can be improved to clarify intent:

```swift
func test_errorView_hidesErrorMessageOnTap() {
    let (sut, loader) = makeSUT()

    sut.loadViewIfNeeded()
    XCTAssertEqual(sut.errorMessage, nil)

    loader.completeFeedLoadingWithError()
    XCTAssertEqual(sut.errorMessage, localized("FEED_VIEW_CONNECTION_ERROR"))

    sut.simulateTapOnErrorMessage()
    XCTAssertEqual(sut.errorMessage, nil)
}
```

**Review by caiozullo — COMMENTED** [review]
Hi @xvladax, you almost passed the challenge!

Please, check my review comments and let me know once you're done so I can review it again. 

Thank you 🙂

**Comment by xvladax** [comment]
Hi @caiozullo,

I applied all suggestions.

Thanx

**Review by caiozullo — COMMENTED** [review]
Hi @xvladax,

Congratulations, you've passed the challenge 🥇

- All tests are passing ✅
- Your code is clean and handles edge cases ✅
- The UI and animations look and behave as expected ✅
- Light and Dark mode support ✅
- Added an extra language ✅

There's only one suggestion about renaming a View Model property that wasn't addressed. Please, check the review comment and feel free to push changes.

Great job!
